fix(core): input coercion

lekhmanrus opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Added common coercion  functions/types (same as in **@angular/cdk/coercion**) that could be used for boolean/number property coercion (with **@Input** mostly).
Currently each package has it's own local functions/expressions for doing coercion. And they are not consistent.
Small example from email validator - forms package:
```html
  <input matInput type="email" name="email" [(ngModel)]="email" email="email" required="required" />
```
`@Input email` is `false` currently.


## What is the new behavior?
If consistent coercion would be applied (same as in **cdk**) - that `@Input email` accessor returns `true`. In addition, consistent coercion could be used across solution, and no need to type custom functions for that (`attrBoolValue` in **packages\router\src\directives\router_link.ts**).


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

lekhmanrus committed:
fix(core): input coercion

BREAKING CHANGE: Forms [email] input coercion

Forms [email] input value will be considered as true if it is defined with any value rather
than false and 'false'.

sod commented on packages/core/src/util/coercion.ts:
```diff
@@ -0,0 +1,42 @@
+
+/** Coerces a data-bound value (typically a string) to a boolean. */
+export function coerceBooleanProperty(value: any): boolean {
+  return value != null && `${value}` !== 'false';
```
For memory reasons and as this is a framework, would it be sensible to do
```ts
return typeof value === 'boolean' ? value : value != null && value !== 'false';
```
 just to prevent wasting memory by creating new strings?

lekhmanrus replied:
Good catch! Thanks!
Probably the same should be done for **@angular/cdk/coercion**.

dario-piotrowicz commented:
Hi @lekhmanrus I don't think your commits follow the commit message standards: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-format

also regarding your last change I believe you should just fix it up into the other one and have a single commit :slightly_smiling_face:

dylhunn commented:
I think this is technically a breaking change, although hopefully not in any substantial sense. I do like the idea at first glance though. We'd need to run a TGP to determine the extent of the breakage.

atscott commented on packages/forms/src/directives/validators.ts:
```diff
@@ -349,12 +349,12 @@ export class RequiredValidator implements Validator {
-  get required(): boolean|string {
+  get required(): BooleanInput {
     return this._required;
   }
 
-  set required(value: boolean|string) {
-    this._required = value != null && value !== false && `${value}` !== 'false';
+  set required(value: BooleanInput) {
+    this._required = coerceBooleanProperty(value);
```
These types are not correct. `this._required` is set to `coerceBooleanProperty` which only returns a `boolean`. The getter should not return `BooleanInput` but `boolean` instead.

Other than that, everything else here looks reasonable.

lekhmanrus replied:
@atscott Thanks! Fixed.

dylhunn reviewed (approved):
reviewed-for: fw-forms, fw-core

atscott reviewed (approved):
reviewed-for: fw-router

dylhunn commented:
This change looks good, but needs a few more things:
- The commit message doesn't match the correct format (see the lint check)
- Breaking change message must be added to commit message
- There are some failing tests
- A TGP will be required (I'll do this after we figure out the failing tests)

lekhmanrus commented:
Hi @dylhunn
* [x] fixed commit message
* [x] added breaking change message
* [x] fixed tests, except `aio`. It's not quite clear what is going wrong with [that](https://app.circleci.com/pipelines/github/angular/angular/42370/workflows/8a89551b-dcb0-4842-97cf-5be94044950d/jobs/1117758?invite=true#step-104-204)?

dylhunn commented:
@lekhmanrus Thanks! I will have a look at the aio failure. It's a [Dgeni](https://github.com/angular/dgeni) error message, indicating something is wrong with a JSDoc comment.

dylhunn commented:
@lekhmanrus I think the docs error is happening because you are using a tag I've never seen: `@docs-private`. Try using just `@private`. (The fact that Dgeni crashes without printing an error actually was fixed a few days ago [in this PR](github.com/angular/dgeni-packages/pull/323), but it's not released yet.)

dylhunn reviewed (approved):
reviewed-for: fw-forms, fw-core, public-api

lekhmanrus commented:
Thanks @dylhunn! 🙂
Interesting, `@docs-private` is using in [angular/components](https://github.com/angular/components/search?q=docs-private).

AndrewKushnir commented on goldens/public-api/forms/forms.md:
```diff
@@ -204,7 +205,8 @@ export class DefaultValueAccessor extends BaseControlValueAccessor implements Co
 export class EmailValidator extends AbstractValidatorDirective {
-    email: boolean | string;
+    get email(): boolean;
+    set email(value: ɵBooleanInput);
```
I think we should not expose new input types in a public API.

My proposal for this PR is to:
- introduce new helper utilities
- use them internally to make the code more consistent (and avoid repeating the code)
- avoid changing public API surface

Once this PR lands, we can re-assess whether introducing new input types are needed.

AndrewKushnir commented on packages/core/src/util/coercion.ts:
```diff
@@ -0,0 +1,42 @@
+export type NumberInput = string|number|null|undefined;
+
+/** Coerces a data-bound value (typically a string) to a number. */
+export function coerceNumberProperty(value: any): number;
```
This function looks unused? If so, we should exclude it from the change and introduce it when there is a need to have it.

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -320,8 +320,7 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
     const disabledValue = changes['isDisabled'].currentValue;
-
-    const isDisabled = disabledValue === '' || (disabledValue && disabledValue !== 'false');
+    const isDisabled = disabledValue !== 0 && coerceBooleanProperty(disabledValue);
```
Could you please add a comment to explain why we also have the `!== 0` check here?

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -374,7 +374,14 @@ export class RequiredValidator extends AbstractValidatorDirective {
    * Tracks changes to the required attribute bound to this directive.
    */
-  @Input() required!: boolean|string;
+  @Input()
+  get required(): boolean {
+    return this._required;
+  }
+  set required(value: BooleanInput) {
+    this._required = coerceBooleanProperty(value);
+  }
+  private _required!: boolean;
```
This coercion happens in the `normalizeInput` function below. Similarly, the `EmailValidator` also has it in the `normalizeInput` function. Note: the logic used in those functions are different for historical reasons. We should try to align them and use the `coerceBooleanProperty` there, but also figure out if there are potential cases when the behavior changes.

AndrewKushnir replied:
Also, I think we should rename the `coerceBooleanProperty` function to drop the `Property` part as it encodes the location at which the function is typically used. However, we can use it in various places. My proposal is to make it more generic (like `coerceToBoolean`).

dylhunn commented:
TAP train rerun looks good: [link](https://fusion2.corp.google.com/presubmit/tap/426498791/OCL:426498791:BASE:426485198:1644043646229:2305c1f9/targets)

dylhunn commented:
We need a couple more public-api approvals on this.

dylhunn reviewed (approved):
reviewed-for: fw-forms, fw-core

dylhunn commented:
This PR was merged into the repository by commit d5719c2e0fb237be71d658444bbfe2410e05086a.

dylhunn commented:
@lekhmanrus, thank you for your contribution to Angular!
